### PR TITLE
Improve documentation on `ColumnarValue`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of #7013 

## Rationale for this change

While reviewing https://github.com/apache/datafusion/pull/10148 I realized that the documentation for `ColumnarValue` could be improved.

## What changes are included in this PR?
Improve the documentation

## Are these changes tested?
CI

## Are there any user-facing changes?
Better docs. No code change

### CODE REVIEW

File: datafusion/expr/src/columnar_value.rs [hunk 1]
```diff
@@ -25,15 +25,65 @@ use datafusion_common::format::DEFAULT_CAST_OPTIONS;
...
 
-/// Represents the result of evaluating an expression: either a single
-/// [`ScalarValue`] or an [`ArrayRef`].
+/// The result of evaluating an expression.
 ///
-/// While a [`ColumnarValue`] can always be converted into an array
-/// for convenience, it is often much more performant to provide an
-/// optimized path for scalar values.
+/// [`ColumnarValue::Scalar`] represents a single value repeated any number of
+/// times. THis is an important performance optimization for handling values
```
Comment (waynexia) [inline datafusion/expr/src/columnar_value.rs:31]: nit:

```suggestion
/// times. This is an important performance optimization for handling values
```

File: datafusion/expr/src/columnar_value.rs [hunk 2]
```diff
@@ -25,15 +25,65 @@ use datafusion_common::format::DEFAULT_CAST_OPTIONS;
...
+/// When implementing functions or operators, it is important to consider the
+/// performance implications of handling scalar values.
+///
+/// Because all functions must handle [`ArrayRef`], it is
+/// convenient to convert [`ColumnarValue::Scalar`]s using
+/// [`Self::into_array`]. For example,  [`ColumnarValue::values_to_arrays`]
+/// converts multiple columnar values into arrays of the same length.
+///
+/// However, it is often much more performant to provide a different,
+/// implementation that handles scalar values differently
```
Comment (waynexia) [inline datafusion/expr/src/columnar_value.rs:86]: 👍

**Review by waynexia — APPROVED** [review]
LGTM, thanks 👍

**Review by comphead — APPROVED** [review]
lgtm thanks @alamb

**Comment by jayzhan211** [comment]
Thanks @alamb @waynexia and @comphead

**Pull request merged**